### ISSUE TITLE
Exception: Unable to resolve symbol: `model` in this context

### ISSUE THREAD

**Issue by BrunoBonacci** [yieldbot/flambo#82]
Hi guys,
we having a strange problem here.

We are running *spark-streaming* to do real-time recommendations. So we train a model based on ALS with batch processes, then we export the model and during the streaming we just 'score' the items based on the trained model.
In order to do so we need to Broadcast the model to the workers, and in the mappers execute the *recommendProducts*.
However when we try to build a mapper function which closes over the broadcast model it seems to have some issues during the 'deserialization' of the locals.

We tried to reduce the code to the shortest possible example and we built a similar streaming job which rather than consuming from Kafka, just receives the stream from a socket port (similar to the hello-world spark streaming example),  so here is code sample

```Clojure
(ns flambo-test.core
  (:require [flambo.conf :as conf])
  (:require [flambo.api :as f])
  (:require [flambo.broadcast :as b]
            [flambo.streaming :as fstream])
  (:import  [org.apache.spark.mllib.recommendation MatrixFactorizationModel]
            [org.apache.spark.api.java JavaSparkContext]
            [org.apache.spark.streaming.dstream ConstantInputDStream]
            [scala.reflect ClassTag$]
            )
  (:gen-class))


(defn spark-config []
  (-> (conf/spark-conf)
      (conf/master "local[*]")
      (conf/set "spark.broadcast.factory" "org.apache.spark.broadcast.HttpBroadcastFactory")
      (conf/app-name "Flambo Test")))


(defn context [config]
  (fstream/streaming-context config 2000))

;;
;; HERE loading the trained model
;; and broadcast it
;;
(defn load-model [context model-path]
  (let [sc (.sparkContext context)]
    (b/broadcast sc
                 (MatrixFactorizationModel/load
                  (JavaSparkContext/toSparkContext sc)
                  model-path))))


(defn to-long [^String s] (Long/parseLong s))

;;
;; HERE SEEMS TO BE THE PROBLEM
;; We are creating a function to map over the stream
;; which closes over the *model* for every 
;; incoming value we run `recommendProducts`
;; and then we will use the output to extract
;; recommended items. However when running this
;; we are getting the following exception
;; Exception: Unable to resolve symbol: `model` in this context
;;
(defn do-model-scoring
  [^org.apache.spark.broadcast.Broadcast model]
  (f/fn [x]
    (let [m (b/value model)]
      (.recommendProducts m (to-long x) 10))))


(defn input-stream [context host port]
  (.socketTextStream context host port))


(defn do-score-stream! [stream b-model]
  (-> stream
      ;; scoring the items 
      (fstream/map (do-model-scoring b-model) )
      ;; just printing the output for testing purposes
      (fstream/print)))
```

Now here is the interesting part:

* When we run the example as follow it seems to be working

```Clojure
  ;; using global vars works every time

  (def ssc (context (spark-config)))

  ;; `model` is defined globally
  (def model (load-model ssc "/tmp/model"))

  (do-score-stream!
   (input-stream ssc "localhost" 9999) model)

  (.start ssc)

  ;; stop it with:
  ;; (.stop ssc)
```

* When the model is in a closure (wraps over a local value) it works on the first item but not on any subsequent item, and the *Exception: Unable to resolve symbol: `model` in this context* is thrown.

```Clojure
  ;; using a local `model` works only the first time

  (def ssc (context (spark-config)))

  ;; here `model` is a local binding.
  (let [model (load-model ssc "/tmp/model")]
    (do-score-stream!
     (input-stream ssc "localhost" 9999) model))

  (.start ssc)

  ;; stop it with:
  ;; (.stop ssc)
```

This is the exception we get in the from the second item onwards
```
java.lang.RuntimeException: Unable to resolve symbol: model in this context, compiling:(NO_SOURCE_PATH:38:13)
	at clojure.lang.Compiler.analyze(Compiler.java:6543)
	at clojure.lang.Compiler.analyze(Compiler.java:6485)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3791)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6725)
	at clojure.lang.Compiler.analyze(Compiler.java:6524)
	at clojure.lang.Compiler.access$300(Compiler.java:38)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6129)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6723)
	at clojure.lang.Compiler.analyze(Compiler.java:6524)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6711)
	at clojure.lang.Compiler.analyze(Compiler.java:6524)
	at clojure.lang.Compiler.analyze(Compiler.java:6485)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:5861)
	at clojure.lang.Compiler$FnMethod.parse(Compiler.java:5296)
	at clojure.lang.Compiler$FnExpr.parse(Compiler.java:3925)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6721)
	at clojure.lang.Compiler.analyze(Compiler.java:6524)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6711)
	at clojure.lang.Compiler.analyze(Compiler.java:6524)
	at clojure.lang.Compiler.analyze(Compiler.java:6485)
	at clojure.lang.Compiler$InvokeExpr.parse(Compiler.java:3791)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6725)
	at clojure.lang.Compiler.analyze(Compiler.java:6524)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6711)
	at clojure.lang.Compiler.analyze(Compiler.java:6524)
	at clojure.lang.Compiler.analyze(Compiler.java:6485)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:5861)
	at clojure.lang.Compiler$LetExpr$Parser.parse(Compiler.java:6179)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6723)
	at clojure.lang.Compiler.analyze(Compiler.java:6524)
	at clojure.lang.Compiler.analyze(Compiler.java:6485)
	at clojure.lang.Compiler$BodyExpr$Parser.parse(Compiler.java:5861)
	at clojure.lang.Compiler$FnMethod.parse(Compiler.java:5296)
	at clojure.lang.Compiler$FnExpr.parse(Compiler.java:3925)
	at clojure.lang.Compiler.analyzeSeq(Compiler.java:6721)
	at clojure.lang.Compiler.analyze(Compiler.java:6524)
	at clojure.lang.Compiler.eval(Compiler.java:6779)
	at clojure.lang.Compiler.eval(Compiler.java:6745)
	at clojure.core$eval.invoke(core.clj:3081)
	at serializable.fn$fn__362$fn__370.invoke(fn.clj:204)
	at serializable.fn$fn__362.invoke(fn.clj:203)
	at clojure.lang.MultiFn.invoke(MultiFn.java:233)
	at serializable.fn$deserialize.invoke(fn.clj:165)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invoke(core.clj:630)
	at clojure.core.memoize$through_STAR_$fn__813.invoke(memoize.clj:66)
	at clojure.core.cache$through$fn__619.invoke(cache.clj:55)
	at clojure.core.memoize$through_STAR_$fn__809$fn__810.invoke(memoize.clj:65)
	at clojure.core.memoize$d_lay$reify__804.deref(memoize.clj:54)
	at clojure.core$deref.invoke(core.clj:2206)
	at clojure.core.memoize$build_memoizer$fn__864.doInvoke(memoize.clj:152)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.AFunction$1.doInvoke(AFunction.java:29)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at flambo.function$Function_call$fn__985.invoke(function.clj:83)
	at flambo.function$Function_call.doInvoke(function.clj:82)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at flambo.function.Function.call(Unknown Source)
	at org.apache.spark.api.java.JavaPairRDD$$anonfun$toScalaFunction$1.apply(JavaPairRDD.scala:1027)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:328)
	at scala.collection.Iterator$$anon$10.next(Iterator.scala:312)
	at scala.collection.Iterator$class.foreach(Iterator.scala:727)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
	at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:48)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:103)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:47)
	at scala.collection.TraversableOnce$class.to(TraversableOnce.scala:273)
	at scala.collection.AbstractIterator.to(Iterator.scala:1157)
	at scala.collection.TraversableOnce$class.toBuffer(TraversableOnce.scala:265)
	at scala.collection.AbstractIterator.toBuffer(Iterator.scala:1157)
	at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:252)
	at scala.collection.AbstractIterator.toArray(Iterator.scala:1157)
	at org.apache.spark.rdd.RDD$$anonfun$take$1$$anonfun$28.apply(RDD.scala:1298)
	at org.apache.spark.rdd.RDD$$anonfun$take$1$$anonfun$28.apply(RDD.scala:1298)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1839)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1839)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:88)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: Unable to resolve symbol: model in this context
	at clojure.lang.Util.runtimeException(Util.java:221)
	at clojure.lang.Compiler.resolveIn(Compiler.java:7019)
	at clojure.lang.Compiler.resolve(Compiler.java:6963)
	at clojure.lang.Compiler.analyzeSymbol(Compiler.java:6924)
	at clojure.lang.Compiler.analyze(Compiler.java:6506)
	... 82 more
```

* Finally if i declare `model` as a global var, it works again

```Clojure
  ;; hack to make it work

  ;; define a global `model'
  (def model nil)
  (def ssc (context (spark-config)))

  (let [m (load-model ssc "/tmp/model")]
    ;; update the global var `model'
    (alter-var-root #'model (constantly m))
    (do-score-stream!
     ;; use the global `model' rather than the local
     (input-stream ssc "localhost" 9999) model))

  (.start ssc)

  ;; stop it with:
  ;; (.stop ssc)
``` 

We are using Flambo 0.6.0 with Spark 1.5.0. We tried Spark 1.4.1 with the same result.
Can anyone help us to fix this issue?

**Comment by sorenmacbeth** [comment]
Hello,

Thanks for the detailed report. The root of the problem is the anything closed over by a serializable function must itself be serializable by kryo. What's probably happening is that kryo is unable to figure out how to serialize sparks Broadcast object. The serializable function library is "best effort" in that it tries to serialize closures with kryo and if it can't it just skips the object. When it tries to deserialize, the object isn't there so you get the error you are seeing. 

The solution to this is to write a kryo serializer/deserializer for Broadcast objects. I can take a look at this soon, but the interface is pretty simple if you wanted to take a crack at it first and send over a PR.

**Comment by sorenmacbeth** [comment]
I released 0.7.1 which should fix this issue with broadcast vars (and accumulators).

**Comment by crankyadmin** [comment]
Superb! I'll try this later.

On Wednesday, 30 September 2015, Soren Macbeth <notifications@github.com>
wrote:

> Closed #82 <https://github.com/yieldbot/flambo/issues/82>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/yieldbot/flambo/issues/82#event-423468076>.
>


-- 
Dave
-------------
Email: root@crankyadmin.net
Web: www.crankyadmin.net
Twitter: @crankyad
